src/Ouroboros/Network/ChainFragment.hs

```diff
@@ -0,0 +1,540 @@
+                 => (block1 -> block2) -> ChainFragment block1 -> ChainFragment block2
+mapChainFragment f (ChainFragment c) = ChainFragment (FT.fmap' f c)
+
+singleton :: HasHeader block => block -> ChainFragment block
+singleton = ChainFragment . FT.singleton
+
+-- TODO more efficient with a fold? -> avoid constructing views
+valid :: HasHeader block => ChainFragment block -> Bool
+valid Empty = True
+valid (c :> b) = valid c && validExtension c b
+
+-- | Checks whether the first block @bSucc@ is a valid successor of the second
+-- block @b@.
+--
+-- * The 'blockPrevHash' of the @bSucc@ must match that of @b@.
+-- * The 'blockSlot' of @bSucc@ must be strictly larger than that of @b@.
+-- * The 'blockNo' of @bSucc@ must be 1 greater than that of @b@.
+--
+-- This function does not check whether any of the two blocks satisfy
+-- 'blockInvariant'.
+isValidSuccessorOf :: HasHeader block
+                   => block  -- ^ @bSucc@
+                   -> block  -- ^ @b@
+                   -> Bool
+isValidSuccessorOf bSucc b =
+    pointHash p == blockPrevHash bSucc
+ && pointSlot p <  blockSlot bSucc
+ && blockNo   b == pred (blockNo bSucc)
+  where
+    p = blockPoint b
+
+validExtension ::  HasHeader block => ChainFragment block -> block -> Bool
+validExtension c bSucc = blockInvariant bSucc && case head c of
+  Nothing -> True
+  Just b  -> bSucc `isValidSuccessorOf` b
+
+head :: HasHeader block => ChainFragment block -> Maybe block
+head (_ :> b) = Just b
+head Empty    = Nothing
+
+headPoint :: HasHeader block => ChainFragment block -> Maybe (Point block)
+headPoint (_ :> b) = Just (blockPoint b)
+headPoint Empty    = Nothing
+
+headSlot :: HasHeader block => ChainFragment block -> Maybe Slot
+headSlot c = pointSlot <$> headPoint c
+
+headHash :: HasHeader block => ChainFragment block -> Maybe (Hash block)
+headHash c = pointHash <$> headPoint c
+
+headBlockNo :: HasHeader block => ChainFragment block -> Maybe BlockNo
+headBlockNo (_ :> b) = Just (blockNo b)
+headBlockNo Empty    = Nothing
+
+
+-- | Make a list of blocks from a 'ChainFragment', in newest-to-oldest order.
+toNewestFirst :: HasHeader block => ChainFragment block -> [block]
+toNewestFirst = foldChainFragment (flip (:)) []
+
+-- | Make a list of blocks from a 'ChainFragment', in oldest-to-newest order.
+toOldestFirst :: HasHeader block => ChainFragment block -> [block]
+toOldestFirst (ChainFragment ft) = Foldable.toList ft
+
+-- | Make a 'ChainFragment' from a list of blocks in newest-to-oldest order.
+fromNewestFirst :: HasHeader block => [block] -> ChainFragment block
+fromNewestFirst bs = foldr (flip (:>)) Empty bs
+
+-- | Make a 'ChainFragment' from a list of blocks in oldest-to-newest order.
+fromOldestFirst :: HasHeader block => [block] -> ChainFragment block
+fromOldestFirst bs = assert (valid c) c
+  where
+    c = ChainFragment $ FT.fromList bs
+
+-- | Drop the last @n@ blocks from the 'ChainFragment'.
+drop :: HasHeader block
```

Comment:
This is perhaps confusingly named. `drop` normally drops from the _head_ of the list, but here it's dropping from the end. Might be good to pick a more descriptive name. `dropNewest` perhaps?

Reply:
Good suggestion, I'll rename it to `dropNewest`.